src/lib/cert/x509/key_constraint.cpp

```diff
@@ -54,11 +25,11 @@ Key_Constraints find_constraints(const Public_Key& pub_key,
       constraints |= KEY_AGREEMENT;
 
    if(name == "RSA" || name == "ElGamal")
-      constraints |= KEY_ENCIPHERMENT | DATA_ENCIPHERMENT;
+      constraints |= KEY_ENCIPHERMENT | DATA_ENCIPHERMENT | ENCIPHER_ONLY | DECIPHER_ONLY;
 
    if(name == "RSA" || name == "RW" || name == "NR" ||
-      name == "DSA" || name == "ECDSA")
-      constraints |= DIGITAL_SIGNATURE | NON_REPUDIATION;
+      name == "DSA" || name == "ECDSA" || name == "ECGDSA" || name == "ECKCDSA")
+      constraints |= DIGITAL_SIGNATURE | NON_REPUDIATION | KEY_CERT_SIGN | CRL_SIGN;
```

Comment:
I would rename it to `match_constraints()` and have it throw an exception if limits contains a constraint that it not allowed for the key type, and return `void` otherwise, so lines 34ff. would become:

```
   if ( ( limits & constraints ) != limits )
      {
      throw Exception("Constraint not permitted for this key type");
      }
```

That would change the behaviour of commands like `gen_self_signed` such that all end entity certificates would contain no constraints, if not explicitly given, which seems correct to me. Would you support such an implementation?